File: readers-writers/readerwriter.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <sys/time.h>

#define RESOURCE_SIZE 50
#define NUM_WRITERS 4
#define NUM_READERS 9

int count;
pthread_mutex_t mx_count;
pthread_mutex_t mx_write;

char resource[RESOURCE_SIZE];


void write_resource(int wid)
{
	printf("w%d-> start\n", wid);

	int i = 0;
	while (i < RESOURCE_SIZE)
	{
		char c = resource[i];
		if (c == '\0') {
			resource[i] = wid + '0';
			break;
		}
		i++;
	}
	
	printf("w%d-> done\n", wid);
}

void read_resource(int rid)
{
	printf("r%d-> read: %s\n", rid, resource);
}

void* writer_sim(void* arg)
{
	int id = *(int*)arg;
	
	pthread_mutex_lock(&mx_write);
	write_resource(id);
	pthread_mutex_unlock(&mx_write);
	
	pthread_exit(0);
}

void* reader_sim(void* arg)
{
	int id = *(int*)arg;
	
	//lock counter mx and increment count
	pthread_mutex_lock(&mx_count);
	count++;
	printf("r%d-> start | r_count: %d\n", id, count);
	if (count == 1)
		pthread_mutex_lock(&mx_write);

	pthread_mutex_unlock(&mx_count);
	
	read_resource(id);
	
	//lock counter mx and decrement count
	pthread_mutex_lock(&mx_count);
	count--;
	printf("r%d-> done\n", id);
	if (count == 0)
		pthread_mutex_unlock(&mx_write);
		
	pthread_mutex_unlock(&mx_count);
		
	pthread_exit(0);
}


int main(int argc, char* argv[])
{
	//Start timer
	struct timeval stop, start;
	gettimeofday(&start, NULL);
	
	//Initialize the resource and mutexes
	int i = 0;
	while (i < RESOURCE_SIZE)
	{
		resource[i] = '\0';
		i++;
	}
	count = 0;
	pthread_mutex_init(&mx_count, NULL);
	pthread_mutex_init(&mx_write, NULL);
	
	//initialize the readers and writers stuff
	pthread_t writers[NUM_WRITERS];
	pthread_t readers[NUM_READERS];

	int wids[NUM_WRITERS] = { 1,2,3,4 };
	int rids[NUM_READERS] = { 1,2,3,4,5,6,7,8,9 };
	int w = 0;
	int r = 0;
	
	//start the writers
	while (w < NUM_WRITERS)
	{
		if (pthread_create(&writers[w], NULL, writer_sim, &wids[w]) != 0)
		{
			printf("Failed to create writer thread!\n");
			exit(EXIT_FAILURE);
		}
		w++;
	}
	
	//start the readers
	while (r < NUM_READERS)
	{
		if (pthread_create(&readers[r], NULL, reader_sim, &rids[r]) != 0)
		{
			printf("Failed to create reader thread!\n");
			exit(EXIT_FAILURE);
		}
		r++;
	}
	
	//wait for writers to finish
	w = 0;
	while (w < NUM_WRITERS) {
		pthread_join(writers[w++], NULL);
	}
	
	//wait for readers to finish
	r = 0;
	while (r < NUM_READERS) {
		pthread_join(readers[r++], NULL);
	}
	
	//destroy the mutexes
	pthread_mutex_destroy(&mx_count);
	pthread_mutex_destroy(&mx_write);
	
	gettimeofday(&stop, NULL);
	printf("Simulation finished. Runtime: %dusec\n", (int)(stop.tv_usec - start.tv_usec));
	return 0;
} 

File: readers-writers/ReaderWriter.java
import java.util.concurrent.*;
import java.util.Random;

public class ReaderWriter
{
	//reader count and resource to be shared
	public static int numReaders;
	public static String resource;
	
	//locks
	public static Semaphore readLock = new Semaphore(1);
	public static Semaphore writeLock = new Semaphore(1);
	
	///Writer thread
	private static class Writer extends Thread
	{
		private Thread thread;
		private int id;
		
		Writer(int id)
		{
			this.id = id;
			
			thread = new Thread(this, "WriterThread[" + id + "]");
		}
		
		public void run()
		{
			try {
				writeLock.acquire();
				System.out.println("w" + id + "-> start");
				resource = resource + id;
				System.out.println("w" + id + "-> done");
				writeLock.release();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
	
	///Reader thread
	private static class Reader extends Thread
	{
		private Thread thread;
		private int id;
		
		Reader(int id)
		{
			this.id = id;
			
			thread = new Thread(this, "ReaderThread[" + id + "]");
		}
		
		public void run()
		{
			try {
				//grab the reader count lock and increment the count
				readLock.acquire();
				numReaders++;
				System.out.println("r" + id + "-> start | r_count: " + numReaders);
				if (numReaders == 1)
					writeLock.acquire();
				readLock.release();
				
				//perform the read
				System.out.println("r" + id + "-> read: " + resource);

				//grab the reader count lock and decrement the count
				readLock.acquire();
				numReaders--;
				System.out.println("r" + id + "-> done");
				if (numReaders == 0)
					writeLock.release();
				readLock.release();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
	}
	
	
	public static void main(String[] args) 
	{
		final int NUM_WRITERS = 4;
		final int NUM_READERS = 9;
		
		//start the timer (msecs)
		long startTime = System.nanoTime() / 1000;
		
		//initializations
		numReaders = 0;
		resource = "";
		Writer[] writers = new Writer[NUM_WRITERS];
		Reader[] readers = new Reader[NUM_READERS];
		
		//start the writers
		for(int i = 0; i < NUM_WRITERS; i++)
		{
			writers[i] = new Writer(i+1); 
			writers[i].start();
		}
		
		//start the readers
		for(int i = 0; i < NUM_READERS; i++)
		{
			readers[i] = new Reader(i+1);
			readers[i].start();
		}

		
		//wait for writers to finish
		for(int i = 0; i < NUM_WRITERS; i++)
		{
			try { writers[i].join(); }
			catch (InterruptedException e) { e.printStackTrace(); }
		}
		
		//wait for readers to finish
		for(int i = 0; i < NUM_READERS; i++)
		{
			try { readers[i].join(); }
			catch (InterruptedException e) { e.printStackTrace(); }
		}
		
		long elapsedTime = (System.nanoTime() / 1000) - startTime;
		System.out.println("Simulation finished. Runtime: " + elapsedTime + "usec");
	}
}
